### PR TITLE
Coursesetting

### PR DESCRIPTION
**Pull request by pauliojanpera**
Hi!

Your project gave me a kick start to implementing some of my ideas. Thank you a lot. =)=)=)

I am working to view Ocad courses in SVG. I fixed some bugs in the code and switched the text label renderer to use HTML div's.  I do not know if you consider that a good approach because it will limit the SVG viewers to those which can view HTML too. Div's provide us multiline rendering for free, though. One possibility would be to render only the multiline texts as divs..

Anyway, there are no brainer bugfixes included.

### CODE REVIEW

**Comment by perliedman** [comment]
Hi,
thanks for this. it looks like there are quite a few changes in the same PR here, I will try to have a look soon, although I'm currently on vacation so can't promise exactly when.

As a side note, I'm working on a similar project called [O-Scout](https://github.com/perliedman/o-scout), although it can currently only read Purple Pen courses. At this moment SVG/PDF is broken, but it's very close to working. Maybe I can get it to read OCAD courses as well with your PR!

File: src/ocad-reader/object-index.js [hunk 1]
```diff
@@ -4,7 +4,12 @@ const TObject = require('./tobject')
 module.exports = class ObjectIndex {
   constructor(reader, version) {
     this.version = version
-    this.nextObjectIndexBlock = reader.readInteger()
+
+    // Ignore pointers that do not point to a valid location in the file.
+    // Compare getBlockCheckedRaw() in Open Orienteering Mapper.
+    this.nextObjectIndexBlock = (i =>
+      i > reader.buffer.length - (256 * 40 + 4) ? 0 : i)(reader.readInteger())
+
```
Comment (perliedman) [inline src/ocad-reader/object-index.js:12]: As mentioned in #20, I took this change and made a slight variation to also add a warning when we do this.

File: src/ocad-reader/index.js [hunk 2]
```diff
@@ -131,4 +131,24 @@ class OcadFile {
       : { x: 0, y: 0, m: 1 }
     return new Crs(scalePar)
   }
+
+  getBounds(projection = v => v) {
```
Comment (perliedman) [inline src/ocad-reader/index.js:135]: Great idea, I have missed this method in for example [ocad-tiler](https://github.com/perliedman/ocad-tiler/)!

File: src/ocad-to-svg.js [hunk 3]
```diff
@@ -241,16 +241,14 @@ const objectToSvg = (options, symbols, object) => {
             node.children = node.children.concat([
               lineToPath(
                 object.coordinates,
-                dbl.dblLeftWidth * 1.5 +
-                  dbl.dblRightWidth * 1.5 +
-                  dbl.dblWidth * 2,
```
Comment (perliedman) [inline src/ocad-to-svg.js:246]: I think I experimented with these to get the Mapbox rendering to look correct, but I realize the SVG rendering works without them, so it's probably some peculiarity with Mapbox and shouldn't be part of the reader. :+1:

File: src/ocad-to-svg.js [hunk 4]
```diff
@@ -149,11 +149,20 @@ const patternToSvg = (colors, s) => {
...
-      : document.createTextNode(n.text)
+  let node
+  if (n.text === undefined) {
+    node = document.createElement(n.type)
+    const xmlnss = Object.entries(n.attrs || []).filter(
+      ([key, _]) => key.indexOf('xmlns') === 0
+    )
+    for (const [ns, uri] of xmlnss) {
+      node.setAttributeNS('http://www.w3.org/2000/xmlns/', ns, uri)
+    }
```
Comment (perliedman) [inline src/ocad-to-svg.js:160]: Just checking so I understand... it looks like attributes starting with `xmlns` will be added both here as well as on line 158, but with different namespaces. Does it matter? Should we filter out `xmlns` attributes so they don't get added a second time?

**Review by perliedman — COMMENTED** [review]
Hi again!

I took some time to review this, and let me just say it looks great! Thanks for taking the time, looks like you found quite a few gotchas that apparently haven't surfaced with the OCAD files I have tried.

The only change I feel a bit reluctant about is the use of HTML `div` inside the SVG output. Although I guess it works since SVG is often viewed in a browser, the output isn't really SVG any more? I'd be more comfortable with fixing whatever issues there are with text with normal SVG elements. Perhaps we can take this commit out of the code and discuss it separately?

I have some detail comments below, but as mentioned above, I'm :+1: to your changes!

File: src/ocad-to-svg.js [hunk 5]
```diff
@@ -149,11 +149,20 @@ const patternToSvg = (colors, s) => {
...
 const createSvgNode = (document, n) => {
-  const node =
-    n.text === undefined
-      ? document.createElementNS('http://www.w3.org/2000/svg', n.type)
-      : document.createTextNode(n.text)
+  let node
+  if (n.text === undefined) {
+    node = document.createElement(n.type)
+    const xmlnss = Object.entries(n.attrs || []).filter(
+      ([key, _]) => key.indexOf('xmlns') === 0
```
Comment (perliedman) [inline src/ocad-to-svg.js:156]:
```suggestion
      ([key, _]) => key.startsWith('xmlns')
```

File: src/ocad-reader/object-index.js (hunk 1 above)
Comment (pauliojanpera) [reply src/ocad-reader/object-index.js:12]: This was the first obstacle I encountered when trying to read an Ocad course setting file. I had to setup the devenv and debug Openorienteering mapper to understand that the next block pointer was just gibberish.

File: src/ocad-to-svg.js (hunk 4 above)
Comment (pauliojanpera) [reply src/ocad-to-svg.js:160]: Oh, I think you're right. That's my mistake. Can you help me, what would be the appropriate workflow to correct this?

` n.attrs &&
    Object.keys(n.attrs)
      .filter(attrName => !attrName.startsWith('xmlns'))
      .forEach(attrName => node.setAttribute(attrName, n.attrs[attrName]))`

**Comment by pauliojanpera** [comment]
I understand if you do not wish to have the HTML dependency on the SVG output. That can be left out. However, I think I'm going to proceed implementing course setting features based on HTML+CSS because my target will be the browser platform anyway. Maybe that could be a command line option or an entirely separate export format?

File: src/ocad-to-svg.js (hunk 4 above)
Comment (perliedman) [reply src/ocad-to-svg.js:160]: Looks good to me!

**Comment by perliedman** [comment]
Yes, an option sounds reasonable as long as the pure SVG output has limitations on text!

**Review by perliedman — CHANGES_REQUESTED** [review]
I noticed that the file `pnpm-lock.yaml` was added to the repo, I presume by mistake? Would be nice if you could remove it (unless it was intentional...)!

**Comment by perliedman** [comment]
If you could make the HTML changes optional, or split that part into a separate PR, I'd be happy to merge this!

**Pull request closed without merging**